### PR TITLE
feat: Implement password validation

### LINKED ISSUE #2338: Add password validation during signup.
**Describe the feature you'd like**
The server has implemented a special feature in the password field. It accepts password of minimum length 8, with at least one number, one special character and one text. This check also needs to be implemented in the android client, to inform the user, if they enter some other things.

Parent Issue: #2319 

**Additional context**
This is the security check-in the client-side.

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by atm1504**
Fixes #2338 

Changes:
* Password must be at least 8 characters long
* Must contain at least one alphabet
* Must contain at least one number
* Must contain at least one special character

Screenshots for the change: 
ezgif com-video-to-gif

### CODE REVIEW

File: app/src/main/java/org/fossasia/susi/ai/helper/CredentialHelper.kt [hunk 1]
```diff
@@ -16,7 +16,7 @@ import java.util.regex.Pattern
  */
 object CredentialHelper {
 
-    private val PASSWORD_PATTERN = Pattern.compile("^.{6,64}$")
+    private val PASSWORD_PATTERN = Pattern.compile("^(?=.*[a-z])(?=.*[@#\$%]).{8,64}\$")
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/CredentialHelper.kt]: Add tests

Comment (atm1504) [reply app/src/main/java/org/fossasia/susi/ai/helper/CredentialHelper.kt]: Can you please explain what you wanted to say through this statement.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/susi/ai/helper/CredentialHelper.kt]: I mean add tests for this logic

File: app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt [hunk 2]
```diff
@@ -63,4 +64,12 @@ class SignUpActivityTest {
...
         onView(withId(R.id.signUp)).check(matches(isDisplayed()))
     }
+
+    @Test
+    fun `testSignupPassword`() {
+        Timber.d("running password test..")
+        val password = "123345602"
+        val result = mActivityRule.activity.passwordTest(password)
+        assertEquals("Pattern did not match", false, result)
+    }
```
Comment (iamareebjamal) [inline app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: This is a unit test, not an instrumentation test

File: app/src/main/java/org/fossasia/susi/ai/signup/SignUpActivity.kt [hunk 3]
```diff
@@ -249,6 +249,10 @@ class SignUpActivity : AppCompatActivity(), ISignUpView {
         signUp.isEnabled = true
     }
 
+    fun passwordTest(password: String): Boolean {
+        return CredentialHelper.isPasswordValid(password)
+    }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/signup/SignUpActivity.kt]: Useless

File: app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt (hunk 2 above)
Comment (atm1504) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: Okk, I will update it

**Comment by atm1504** [comment]
@iamareebjamal please review it.

**Review by iamareebjamal — CHANGES_REQUESTED** [review]
Write a unit test. This is not an integration test. Do you want this to purposefully run slower and on Android device when it can run on your PC?

File: app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt [hunk 4]
```diff
@@ -63,4 +64,12 @@ class SignUpActivityTest {
         onView(withId(R.id.signUp)).perform(scrollTo())
         onView(withId(R.id.signUp)).check(matches(isDisplayed()))
     }
+
+    @Test
```
Comment (atm1504) [inline app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: @iamareebjamal the unit test has been added here.

Comment (iamareebjamal) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: Read what's a unit test and what's an instrumentation and integration test

Comment (iamareebjamal) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: Are you testing UI? Then why have you written an Android Instrumentation test which is 10000 times slower than a unit test

Comment (iamareebjamal) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: I wrote the same review 12 days ago https://github.com/fossasia/susi_android/pull/2339/files#r310375934

Seems like you don't take reviews seriously at all

Comment (atm1504) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: Sir, I took that. But I was asked to implement the planned actions first, after that I should resume my task with other PR. This was said to me at the meeting. So, I started working on planned actions, before further updating this pr. After the task on that feature has been implemented, I started working on updating this pr.

Comment (iamareebjamal) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: So then why you asked for review again without changing anything?

Comment (atm1504) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: Sir, in the first review u asked to implement a test. I implemented the unit test. But then you said to me to implement an instrumentation test, which I added today. But, then today in the review u said to remove the instrumentation test. So, the changes have nullified now.
Unit test has already been implemented last week.

Comment (iamareebjamal) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: No, I did not ask you to implement an instrumentation test. You already wrote an instrumentation test. You never wrote a unit test. 

`This is a unit test, not an instrumentation test`

Meant this is a unit test, not an instrumentation test. Why is it in instrumentation test package?

Comment (atm1504) [reply app/src/androidTest/java/org/fossasia/susi/ai/signup/SignUpActivityTest.kt]: That was a misunderstanding. I am rectifying the code. Sorry for the inconvenience caused.

**Comment by atm1504** [comment]
@iamareebjamal I have implemented the unit test. It's present in the test folder.

File: app/src/test/java/org/fossasia/susi/ai/helper/CredentialHelperTest.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+class CredentialHelperTest {
+    @Test
+    fun testSignupPassword() {
+        Timber.d("running password test..")
+        Assert.assertEquals(false, CredentialHelper.isPasswordValid("123345602"))
+        Assert.assertEquals(false, CredentialHelper.isPasswordValid("123345602@"))
+        Assert.assertEquals(false, CredentialHelper.isPasswordValid("123345602a"))
+        Assert.assertEquals(false, CredentialHelper.isPasswordValid("123345602@a"))
+        Assert.assertEquals(true, CredentialHelper.isPasswordValid("123345602@As"))
```
Comment (iamareebjamal) [inline app/src/test/java/org/fossasia/susi/ai/helper/CredentialHelperTest.kt]: Use assertFalse and assertTrue. Remove Timber log

Comment (atm1504) [reply app/src/test/java/org/fossasia/susi/ai/helper/CredentialHelperTest.kt]: Made the suggested changes

**Review by iamareebjamal — APPROVED** [review]

**Pull request merged**